### DISCUSSION TITLE
Unable to patch initContainer onto inject-policy for OPA Envoy sidecar container

### DISCUSSION THREAD

**Discussion by bvamshidhar** [open-policy-agent/community#615] [Community]
I'm trying to add an initContainer add it to the spec using patch operation in the [inject-policy.yaml](https://github.com/open-policy-agent/opa-envoy-plugin/blob/main/examples/istio/quick_start.yaml). I'm adding this initContainer to fetch the data of the other 2 configmaps opa-isito-config and opa-policy from a files stored in a dir /config.
 
For example: opa-istio-config.yaml contains
plugins:
      envoy_ext_authz_grpc:
        addr: :9191
        path: istio/authz/allow
    decision_logs:
      console: true
      
The main purpose of this is to to deploy these two configmaps in opa-istio namespace instead of default namespace.

patch = [{
      "op": "add",
      "path": "/spec/initContainers/-",
      "value": init_container,
    }, {
      "op": "add",
      "path": "/spec/containers/-",
      "value": opa_container,
    }, {
      "op": "add",
      "path": "/spec/volumes/-",
      "value": new_volumes,
    }]

    init_container = {
      "name": "config-fetch"
      "image": "busybox"
      "command": [
        "wget",
        "-P",
        "/config-dir",
        "https://config/opa-istio-config.yaml";,
        "https://config/policy.rego";,
      ],
      "volumeMounts":[{
        "name": "configdir",
        "mountPath": "/config-dir",
      }]
    }
    
    opa_container{
    ...
    }
    
    # I removed the individual volumes for the configmap and added a new one
    new_volumes {
     ...
    }
   
  
   While adding the initContainer, I get an error:
   Error creating: Internal error occurred: add operation does not apply: doc is missing path: "/spec/initContainers/-": missing value
   
   Can you please help me on this issue? 
   Thank you.

**Comment by charlieegan3** [comment]
I think you will need to patch the spec taking into account any existing init containers and including them in the patch if there are any set.

**Reply by bvamshidhar** [reply]
When I previously deployed this, I did not see an initContainer. May I please have the documentation if available?

**Reply by charlieegan3** [reply]
This is an example which shows you how to automatically inject OPA when using Istio as a sidecar container.

If you are interested in injecting an init container, you'll need to do some extra work to make sure that the initContainers field is set before adding to it.

Before you go too far down this road, are you certain that an init container is the best way to go to load the policy into OPA or to make it available? Have you considered using the bundle service API to load the policy into OPA once it starts up?

https://www.openpolicyagent.org/docs/latest/management-bundles/

**Reply by bvamshidhar** [reply]
I will be using only one rego policy which is not going to be changed in the future, I have moved forward with configmap rather than bundle. 
So, how do I set the initContainer field? Does it take it as an attribute?

**Reply by charlieegan3** [reply]
If your Rego policy code is relatively static and is stored in a YAML file in a config map, why not just load the config map as a volume into the OPA container?

**Reply by bvamshidhar** [reply]
I'm dowloading only the data part of the code for opa-istio-config and opa-policy configmaps from a cloud storage, hence the initContainer and also creating a shared volume for this.

**Reply by charlieegan3** [reply]
So you have some of the policy being loaded from cloud storage, and some loaded from a config map?

**Reply by bvamshidhar** [reply]
Yes, I'm trying to load the opa-isito-config and opa-policy from cloud so as to remove both the files from the default namespace. I left the inject-policy configmap as is.

**Reply by charlieegan3** [reply]
Right I see, and just so that I understand, is this because they're different in different environments? Or for some other reason? It seems like it would be easier to load them from configmaps.

**Reply by bvamshidhar** [reply]
If they are in a different namspace, I need to deploy them once in every namespace. I have multiple namespaces. I would be easier to deploy and maintain, if all files could be in one namespace.

**Reply by charlieegan3** [reply]
Perhaps something like https://github.com/mittwald/kubernetes-replicator would be easier?

OPA is designed to load policy from http endpoints when policy is bundled, rather than plain files. You can also use the [Discovery API](https://www.openpolicyagent.org/docs/latest/management-discovery/) to load in config for the OPA itself from a remote bundle. That might be something to consider too.
